### ISSUE TITLE
Feature: Blur Support

### ISSUE THREAD

**Issue by ozwaldorf** [YaLTeR/niri#54]
Would love to see blur support included for transparent windows, likely Dual Kawase, and a nice to have would be a configurable noise pass to mitigate banding from the blur algorithm.

**Comment by phisch** [comment]
Ideally, support the [kde-blur protocol](https://wayland.app/protocols/kde-blur).

**Comment by AyushmanOfficial** [comment]
Any updates?

**Comment by phisch** [comment]
If there were updates, they would be here...

**Comment by ThatOneCalculator** [comment]
Especially now that we have shadows!

**Comment by partisani** [comment]
# i need this, like, right now

<sup>unless you are YaLTeR, in which case i could wait a year or two...</sup>

<sub>also, isn't *[ozwaldorf](https://github.com/ozwaldorf)* the creator of `lutgen-rs`?</sub>

**Comment by AyushmanOfficial** [comment]
> # i need this, like, right now
> 
> unless you are YaLTeR, in which case i could wait a year or two... also, isn't _[ozwaldorf](https://github.com/ozwaldorf)_ the creator of `lutgen-rs`?

Yeah and he also gave the internet this: https://gist.github.com/ozwaldorf/a15539b297a7862af4f12ce07fee6bb7

**Comment by partisani** [comment]
> Yeah and he also gave the internet this: https://gist.github.com/ozwaldorf/a15539b297a7862af4f12ce07fee6bb7

Literally incredible.

**Comment by NTBBloodbath** [comment]
Hi, could you please lock this issue? I'd like to avoid receiving off-topic notifications and more so for a feature request that interests me a lot, and I'm sure I'm not the only one 🙂.

**Comment by YaLTeR** [comment]
I will have to lock if there's any more off topic messages

**Comment by partisani** [comment]
Sorry for that...

**Comment by srjheam** [comment]
Hello,

any progress on this so far?

I've been a Niri user for the past few weeks (I love it btw —it's the best WM I've ever used, and the infinite strip concept is fantastic!), and I'd love to contribute to this feature. I might also use this work for an assignment in my university course.

If any of you have suggestions on where to start with compositors or window manager effects, it would be greatly appreciated.

**Comment by YaLTeR** [comment]
No progress yet and I am not sure this is a good first issue so to say, it needs quite a bit of understanding of the compositor code.

Don't worry I'll get to it eventually.

**Comment by ProsperousTroll** [comment]
Can i haz blur plz?

**Comment by RazoBeckett** [comment]
any updates ?

**Comment by ozwaldorf** [comment]
@YaLTeR Might be worth locking this thread from new comments until you have an update, or someone opens a pr that targets this issue. The "can i haz blur" is only funny so many times 😄

**Comment by Fireye04** [comment]
Looks like there's a semi-working (as far as I can tell) PR here that you can patch if you're desperate: #1634 

Otherwise, you're not going to gain much by bumping an old discussion. It'll happen when it happens. If you actually want this to come faster, you can contribute yourself, or start a bounty.

**Comment by psi4j** [comment]
> Looks like there's a semi-working (as far as I can tell) PR here that you can patch if you're desperate: [#1634](https://github.com/YaLTeR/niri/pull/1634)
> 
> Otherwise, you're not going to gain much by bumping an old discussion. It'll happen when it happens. If you actually want this to come faster, you can contribute yourself, or start a bounty.

If you want to have reasonable latency with the window blur when spawning new windows, it appears this might require some nontrivial architectural changes to the way rendering currently works inside Niri. I tried to do it and realized I don't know enough about the compositor to make those changes without potentially introducing regressions, so I'm leaving it to the pros. 

I hope somebody else here has both the knowledge/skill-set and the interest in making this work while retaining Niri's buttery smooth animations and workflow.

**Comment by b2smile** [comment]
Использую niri неделю, очень понравилось окружение. Хотелось бы, чтобы поддержка размытия фона тоже появилась.

**Comment by CodedNil** [comment]
Would it be an option for this after reading about the complexity involved in implementing blur, to implement something like windows 11 mica as an alternative option to a true blur? Where the compositor creates a heavily blurred image of the wallpaper once then uses that behind any transparent elements, so it looks like you are peeking through a blur into the wallpaper, this would work for most usecases in niri since typically windows do not overlap, it only wouldn't support floating windows.
Ignore me if this only further increases complexity but I'd imagine the performance hit of that is negligible compared to a live blur, and could be used successfully in 90% of situations.

**Comment by psi4j** [comment]
> Where the compositor creates a heavily blurred image of the wallpaper once then uses that behind any transparent elements, so it looks like you are peeking through a blur into the wallpaper,

I think you're describing x-ray blur. That feature already works in fht-compositor's implementation and I think might still be being worked on [here](https://github.com/YaLTeR/niri/pull/1634).

What you can do for now is manually blur a wallpaper using gimp and throw it up as a wallpaper. I've just accepted carefully-used transparency without blur for the time being.

**Comment by Sempyos** [comment]
If you want a scriptable version, I would suggest using ``magick``

**Comment by RazoBeckett** [comment]
is there blur now ? or we still need workarounds and waiting ?

**Comment by ozwaldorf** [comment]
> is there blur now ? or we still need workarounds and waiting ?

If there was, the issue would be closed as completed. Patience, and avoid tagging everyone for no reason 🙏

**Comment by ple1n** [comment]
My app, [an offline instant prefix fuzzy search dictionary completely authored by me, including the algorithm](https://github.com/ple1n/hoverpanel/), uses `kde-blur`, `layershell`, and `data-control`.

I started using Niri because a recent Kwin update broke `data-control` so my app isn't picking up primary selection live.

**Comment by midischwarz12** [comment]
Just trying to give more attention to this PR:
https://github.com/YaLTeR/niri/pull/1634

It doesn't implement the kde-blur protocol which would require apps to use that specific protocol. However, it blurs the background behind windows so any window implementing alpha / transparency can readily support blur.

Hopefullly linking the PR to this issue will put more eyes on it and help development / testing.

**Comment by Sempyos** [comment]
The PR is at a standstill for the moment, I would highly advise you to read the PR for the reason why

**Comment by midischwarz12** [comment]
@Sempyos Are you referring to yalter's work bandwidth? I'd appreciate sending a link to the comment instead of just saying to read the PR.

**Comment by Fireye04** [comment]
> [@Sempyos](https://github.com/Sempyos) Are you referring to yalter's work bandwidth? I'd appreciate sending a link to the comment instead of just saying to read the PR.

They do have a point; a couple links aren't going to convey the full relevant context here. If you do care about this, more than just as a place to throw a couple comments, I'd definitely recommend reading the thing that you're tying to talk about. As I said above, you're not going to gain much by just bumping an old discussion.

That said, other than the couple lingering bugs that I've seen mentioned, the issue seems to be a performance & implementation one. Specifically, here (https://github.com/YaLTeR/niri/pull/1634#issuecomment-2899885885) and here (https://github.com/YaLTeR/niri/pull/1634#issuecomment-3519939263). 

This is a big addition, and is gonna need someone very familiar with both rust and the Niri codebase. Even then, YaLTeR would need to review and modify it, as the change will be a deep one. The original PR author has done great work getting such a demo just working, and they've even been keeping it rebased off main (yay!), but has also stated that they won't be working on it anymore due to the scale of the change here: (https://github.com/YaLTeR/niri/pull/1634#issuecomment-2996715864) and here (https://github.com/YaLTeR/niri/pull/1634#issuecomment-3551847879)

Plainly put, it's very likely going to be awhile. In the meantime, there has been a fork discussed in the thread here: (https://github.com/YaLTeR/niri/pull/1634#issuecomment-3570609024), and as always you're welcome to compile the PR yourself.

**Comment by psi4j** [comment]
> @Sempyos Are you referring to yalter's work bandwidth? I'd appreciate sending a link to the comment instead of just saying to read the PR.

What I understand is that this requires upstream work in Smithay because implementing this properly will require a nontrivial change in the rendering pipeline. A Hyprland style blur implementation is not a viable option for Smithay-based compositors because of the way Smithay isolates components.

Trust me, I would also love to see this implemented sooner rather than later, but I wouldn't hold my breath.

**Comment by midischwarz12** [comment]
@Fireye04 Thank you for providing context. You made it all very clear.

**Comment by oni-swr** [comment]
to quote from #1634 
[comment ](https://github.com/niri-wm/niri/pull/1634#issuecomment-3923676847)
> Closing this in favor of the work being done in https://github.com/niri-wm/niri/tree/wip/branch which is already so much better than what we had here!
> 
> Thank you all folks for trying out my code during these months <3


just an update so you dont have to go through that pr.  apparently its on the wip branch now.

**Comment by YaLTeR** [comment]
Hey everyone, there's a PR now, please test: https://github.com/niri-wm/niri/pull/3483

**Comment by YaLTeR** [comment]
Blur is now merged to main!

**Comment by ozwaldorf** [comment]
Nice work @YaLTeR ! I'll have to give it a try sometime soon

**Comment by yvvki** [comment]
Thank you YaLTeR! Now I can peek the anime girl on the background behind a frosted terminal like a creep :3

**Comment by LBCrion** [comment]
@YaLTeR, I'm working on implementing support for this protocol in sfwbar and wanted to check if my thinking is correct.  I was looking at the case of a borderless window with rounded corners. The shape of the blur is specified via wl_region which uses surface logical coordinates and as a consequence, doesn't allow hidpi shape definition. As a result,  the blur only window will have jagged edges compared to windows with borders on hidpi displays. Is this the case,  or am I missing something?

**Comment by YaLTeR** [comment]
That is correct, yes (and a bit unfortunate). You can hide it underneath an opaque border fairly well.

**Comment by LBCrion** [comment]
The border does indeed cover it nicely. I suspect we'll need a thicker
border for higher scaling factors, but then I've never seen a scaling
factor > 2 in the wild.
